client/asset/eth/eth.go

```diff
@@ -1022,78 +1067,102 @@ func (w *AssetWallet) approvalGas(newGas *big.Int, cfg *dex.Asset) (uint64, erro
 
 // ReturnCoins unlocks coins. This would be necessary in the case of a
 // canceled order.
-func (w *AssetWallet) ReturnCoins(coins asset.Coins) error {
+func (w *ETHWallet) ReturnCoins(coins asset.Coins) error {
+	var amt uint64
+	for _, ci := range coins {
+		c, is := ci.(*fundingCoin)
+		if !is {
+			return fmt.Errorf("unknown coin type %T", c)
+		}
+		if c.addr != w.addr {
+			return fmt.Errorf("coin is not funded by this wallet. coin address %s != our address %s", c.addr, w.addr)
+		}
+		amt += c.amt
+
+	}
+	w.lockedFunds.mtx.Lock()
+	w.unlockFunds(amt, initiationReserve)
+	w.lockedFunds.mtx.Unlock()
+	return nil
+}
+
+// ReturnCoins unlocks coins. This would be necessary in the case of a
+// canceled order.
+func (w *TokenWallet) ReturnCoins(coins asset.Coins) error {
 	var amt, fees uint64
 	for _, ci := range coins {
-		amt += ci.Value()
-		var addr common.Address
-		switch c := ci.(type) {
-		case *fundingCoin:
-			addr = c.addr
-		case *tokenFundingCoin:
-			fees += c.fees
-			addr = c.addr
-		default:
+		c, is := ci.(*tokenFundingCoin)
+		if !is {
 			return fmt.Errorf("unknown coin type %T", c)
 		}
-		if addr != w.addr {
-			return fmt.Errorf("coin is not funded by this wallet. coin address %s != our address %s", addr, w.addr)
+		if c.addr != w.addr {
+			return fmt.Errorf("coin is not funded by this wallet. coin address %s != our address %s", c.addr, w.addr)
 		}
+		amt += c.amt
+		fees += c.fees
 	}
 	if fees > 0 {
-		if w.token == nil {
-			return fmt.Errorf("tokenFundingCoin returned for non-token asset")
-		}
-		w.token.parent.unlockFunds(fees, initiationReserve)
+		ethWallet := w.parent
+		ethWallet.lockedFunds.mtx.Lock()
+		w.parent.unlockFunds(fees, initiationReserve)
```

Comment:
Why not use `w.parent` when accessing the mutex too, and eliminate the `ethWallet` definition?